### DISCUSSION TITLE
Support for private git submodules

### DISCUSSION THREAD

**Discussion by goenning** [vercel/community#44] [Idea]
Does Vercel supports git submodules? I tried a custom build step with `git submodule init && git submodule update` but apparently the build happens outside a git repo

**Comment by goenning** [comment]
Found an alternative solution: use GitHub actions and call Vercel CLI from there. Apparently the automatic git deployment from Vercel does not support Submodules

**Reply by marco-koenen** [reply]
Any update on this? I have the same problem.

**Comment by Coriou** [comment]
Sorry for hijacking this discussion, but I was wondering if this also was working with Serverless Functions ? I have this node function which *requires* a file from a Git submodule directory. While working fine locally, through `vercel dev`, once deployed the function throws an `ENOENT` error because it can not find the file.

I've tried including said file via `vercel.json`'s `includeFiles` config which didn't help.

Works as expected with a file which **isn't** sourced from a submodule directory.

**Reply by Coriou** [reply]
I've actually found the problem. I originally had the submodule's url as a SSH address. I've switched to HTTP and it now works as expected...

**Reply by cnerylozada** [reply]
Thank u so much, It works !

**Comment by styfle** [comment]
We support **public** git submodules over **HTTP** – No need for additional configuration or build commands.

However, we don't have plans to add private submodules or SSH submodules at this time, but maybe in the future.

[See documentation](https://vercel.com/docs/concepts/deployments/build-features#git-submodules)

**Reply by Coriou** [reply]
Thanks. I hadn't seen that documentation, thanks for pointing it out

**Reply by colinwilson** [reply]
> In the future, we'll add support for private submodules 

Yay! I Look forward to this.

**Reply by dillionmegida** [reply]
Looking forward to private submodules 😊

**Reply by tincho** [reply]
arrived here after googling about this + private submodules :+1:

**Reply by JannikZed** [reply]
I get so many times:  
`Warning: Failed to fetch one or more git submodules`
without any description on why it fails. Everything is public and available on github..

**Reply by k-lusine** [reply]
Hi @styfle.

Any update on private submodules?

**Reply by mmskvortsov** [reply]
Hello there. Any news about private submodules integration?

**Reply by beeinger** [reply]
Until Vercel adds support for private submodules, here is a solution:
https://github.com/beeinger/vercel-private-submodule

**Reply by willemmulder** [reply]
Any update on this @styfle?

**Reply by btakita** [reply]
@styfle Please support git submodules using the `git@` protocol. This is causing issues with me setting the root directory setting in a monorepo using git submodules. I cannot use the root directory setting as I need to run a custom script in the monorepo directory to load the submodules...one of which is the website.

**Comment by marcofranssen** [comment]
What about the support of ssh style submodules? Currently I'm facing the following error 

```
WARN: failed to fetch one or more git submodules: Failed to clone 'my-submodule' a second time, aborting
--
```

**Comment by unitof** [comment]
To clarify what @styfle said earlier and after a bit’o frustrating debugging of my own:

Vercel supports Git submodules but **only** cloning them via HTTP or HTTPS, not SSH which is the default. Switch the protocol in your `.gitmodules` from SSH syntax to HTTPS:

Will cause `WARN: failed to fetch one or more git submodules`:
```
[submodule "content"]
	path = content
	url = git@github.com:unitof/self-docs.git
```

Will work hunky dory:
```
[submodule "content"]
	path = content
	url = https://github.com/unitof/self-docs.git
```

**Reply by jessegavin** [reply]
"Will work hunky dory" _IF_ the https url points to a public repo.

**Comment by beeinger** [comment]
Until Vercel adds support for private submodules, here is a solution:
https://github.com/beeinger/vercel-private-submodule

**Reply by alexkushnarov** [reply]
@beeinger Thanks for the solution but it is didn't work for me quite well, for some reason "vercel-install" script was not executing and submodule was not fetched. I solved this by replace "vercel-install" with  "preinstall": "./vercel-submodule-workaround.sh" script.

**Reply by beeinger** [reply]
Oh that is because in Vercel settings (online) you need to change the default installation script from ```install``` to ```vercel-install```

**Reply by alexkushnarov** [reply]
Ok, Thanks!

**Reply by zivc** [reply]
@alexkushnarov your pre-install solution sounds ideal. Before I found this comment I just butchered `vercel-submodule-workaround.sh` to my needs

Instead of:
```
git fetch --depth=1 origin $COMMIT # fetch only the required version
git checkout $COMMIT # checkout on the right commit
```

I used:
```
git fetch origin
git checkout master

npm install
```

Hope this helps anyone else.

**Comment by oscarsala** [comment]
is this still not supported?

**Comment by junhoyeo** [comment]
I made a CLI `vercel-submodules` that clones submodules when added to `Install Command`.

```bash
npx vercel-submodules --all
npx vercel-submodules --paths specific-module-a specific-module-b
```

No need to add workflows or paths one by one; you can leave it to clone all submodules using the `--all` flag, or only clone specific modules with `--paths` (e.g., for ignoring script-only code).
It works for monorepos, ensuring consistent behavior regardless of the *Root Directory* option.
Hope this will be a temporary workaround 🙏 

https://github.com/junhoyeo/vercel-submodules

**Reply by Pagebakers** [reply]
Works perfect, finally able to deploy my repo with private submodules to vercel.

**Reply by junhoyeo** [reply]
> I made a CLI `vercel-submodules` that clones submodules when added to `Install Command`.
> https://github.com/junhoyeo/vercel-submodules

Bump - Thanks to our contributors @harumaxy and @dikshantrajput, we now support [fine-grained PATs](https://github.com/junhoyeo/vercel-submodules#set-github-token)(actually this update was in December) and custom branch support(shipped today) that comes right out of the box!

**Reply by jlorezz** [reply]
@junhoyeo is there a way to also install the node_modules for the respective git submodules? or would we need to build a separate script for this?

**Reply by junhoyeo** [reply]
@florianbress Late response, sorry. [`vercel-submodules`](https://github.com/junhoyeo/vercel-submodules) don't have that feature yet, contributions welcomed(maybe we can manage configs to add a hook to each module after installation).
I would recommend using Yarn Workspaces (but for this you should add your submodules as a fork with `yarn.lock` removed) or `experimental.externalDir` (if you're using Next.js) for now.

**Comment by jessegavin** [comment]
I am chiming in to say that this is a very important feature for my team.

While I am thankful for the creative solution @junhoyeo has made available, it doesn't feel great to hand over a Github token with wide permissions to a 3rd party npm package.

**Comment by smorimoto** [comment]
This workaround allows you to check out multiple submodules using a [fine-grained personal access token](https://github.blog/2022-10-18-introducing-fine-grained-personal-access-tokens-for-github/). If someone needs something like that with absolutely zero dependencies, just take a look.
https://github.com/smorimoto/vercel-private-submodule

**Reply by jessegavin** [reply]
Thank you @smorimoto .  So to use this, I would include the shell script at the root of my project and configure the build settings like so?

![CleanShot 2023-02-10 at 14 56 26@2x](https://user-images.githubusercontent.com/78791/218196023-806208a5-3278-4d43-b81d-9cd2428efd35.png)

**Reply by jessegavin** [reply]
I suppose I would also need to add the environment variable.
![CleanShot 2023-02-10 at 14 59 57@2x](https://user-images.githubusercontent.com/78791/218196523-437de15d-6c48-4584-8998-bd4471b85ae5.png)

(I am posting these images in case they are helpful to others, and in case I am incorrect, in which case, I could learn what I am doing wrong).

**Reply by smorimoto** [reply]
@goenning Yes. It rewrites your submodule configuration from ssh to https and then takes care to set the token to access with https. (You may need to tweak it for your own use, since it's a case of using it as-is. The sed part...)
It also automatically remove your token from the runner if the submodule is checked out or the process fails.

Note: `yarn --immutable` is a feature of yarn v2, so don't forget to configure it for your tool.

**Reply by smorimoto** [reply]
@goenning Good point, indeed. I will update the documentation.

**Reply by smorimoto** [reply]
@goenning Well, whether it should be placed in the root of the project depends on your Vercel setup. If the base directory has changed, it may looks to be higher than the root from that page. (Even if it's actually just specifying the root.)

**Reply by smorimoto** [reply]
If that works, you should get a log like this. Feel free to get in touch if you encounter any bumps.
<img width="972" alt="Screenshot 2023-02-11 at 6 22 50" src="https://user-images.githubusercontent.com/38746192/218200238-fa4202a2-193e-4893-b7ef-18e2a52864b8.png">

**Reply by meatjam** [reply]
thx,this works perfectly and gracefully!

**Comment by ilia-os** [comment]
Please add ssh submodule support. This is really important for all monorepos utilizing submodules.

**Reply by smorimoto** [reply]
FYI: https://github.com/vercel-contrib/vercel-workaround

**Reply by smorimoto** [reply]
It supports SSH sub modules

**Comment by 52617365** [comment]
Jesus this still does not work. Running into `Warning: Failed to fetch one or more git submodules` with a public repo that has a HTTPS url.
Can't seem to find a way to deploy a Hugo site that has a theme set as a git submodule.

**Reply by jessegavin** [reply]
Fetching [public HTTP submodules should work](https://vercel.com/docs/concepts/deployments/build-features#git-submodules) and if it doesn't, it should be filed in a different issue.

This issue is only about supporting private git submodules.

**Reply by smorimoto** [reply]
In the sense of pushing the hack in one place, you should be able to solve it by patching this script
https://github.com/vercel-contrib/vercel-workaround

**Reply by smorimoto** [reply]
Of course I recommend the SSH format, but even if it’s in HTTPS format, it should work perfectly just by patching the line of sed stuff 🙂

**Comment by willemmulder** [comment]
Anything for this on the roadmap yet?

**Reply by beeinger** [reply]
have a look at [this](https://github.com/beeinger/vercel-private-submodule#-approach-1---gitmodules) - if you use github tokens in `.gitmodules` everything works like if the support was there

**Reply by smorimoto** [reply]
If you want to be more polite: https://github.com/vercel-contrib/vercel-workaround/blob/main/vercel-submodule-workaround.sh

**Reply by willemmulder** [reply]
Yeah that works! Still native support would be nice; hence the question 😊
